pre-docs/writing_plans.md

```diff
@@ -88,6 +88,45 @@ or a comma seperated string of target names.
 bolt plan run mymodule::myplan --modulepath ./PATH/TO/MODULES --params load_balancer=lb.myorg.com frontends='["kermit.myorg.com","gonzo.myorg.com"]' backends=waldorf.myorg.com,statler.myorg.com
 ```
 
+## Returning results from plans
+
+> Plans can optionally return a result that can be used from other plans or
+> saved externally.
+
+Plans, unlike functions, are primarily run for side effects but they can
+optionally return a result. To return a result from a plan use the `return`
+function. Any plan that does not call the `return` function will return
+`undef`.
+
+```puppet
+plan return_result(
+  $nodes
+) {
+  return run_task('mytask', $nodes)
+}
+```
+
+The result of a plan must match the `PlanResult` type alias. This roughly
+includes JSON types as well as the Plan lanuguage types which have well defined
+json representations in Bolt.
+
+- `undef`
+- `String`
+- `Numeric`
+- `Boolean`
+- `Target`
+- `Result`
+- `ResultSet`
+- `Error`
+- `Array` with only `PlanResult`
+- `Hash` with `String` keys and `PlanResult` values
+
+or
+
+```
+Variant[Data, String, Numeric, Boolean, Error, Result, ResultSet, Target, Array[Boltlib::PlanResult], Hash[String, Boltlib::PlanResult]]
```

Comment:
Based on https://puppet.com/docs/puppet/5.3/lang_data_abstract.html#data, I would expect `Data` to match scalars.